test/api.test.ts

```diff
@@ -0,0 +1,448 @@
+
+describe('APIs', () => {
+  let server: Server = app.listen(3000);
+  let request: supertest.SuperTest<supertest.Test>;
+
+  before(() => {
+    app.proxy = true;
+    request = supertest(server);
+  });
+
+  describe('Home', () => {
+    it('[GET] /', () => {
+      return request.get('/').expect(200);
+    });
+  });
+
+  describe('Account', () => {
+    const password = '123456';
+
+    it('[POST] /account/signup failed with name', () => {
+      return request
+        .post('/account/signup')
+        .send({ name: '', email, password })
+        .expect(400);
+    });
+
+    it('[POST] /account/signup success', () => {
+      return request
+        .post('/account/signup')
+        .send({ name, email, password })
+        .expect(201);
+    });
+
+    it('[POST] /account/signin success', () => {
+      return request
+        .post('/account/signin')
+        .send({ email, password })
+        .expect(200)
+        .then((response) => {
+          assert(response.headers['set-cookie']);
+        });
+    });
+
+    it('[POST] /account/signin with email not found', async () => {
+      return request
+        .post('/account/signin')
+        .send({ email: 'not-found@email', password })
+        .expect(404);
+    });
+
+    after(async () => {
+      const model = getInstance<UserModel>('userModel');
+      await model.delete({ email });
+    });
+  });
+
+  describe('User', () => {
+    const password = 'mypassword';
+    const persons: {
+      id?: string;
+      email: string;
+      name: string;
+      location: [number, number];
+      password?: string;
+    }[] = [
+      {
+        email,
+        name,
+        password,
+        location: [3, 3]
+      },
+      {
+        email: 'test1@apis.test',
+        name: 'api test1',
+        password,
+        location: [0, 0]
+      },
+      {
+        email: 'test2@apis.test',
+        name: 'api test2',
+        password,
+        location: [10, 10]
+      }
+    ];
+
+    before(async () => {
+      // clear user
+      const userModel = getInstance<UserModel>('userModel');
+      await userModel.delete({});
+
+      const service = getInstance<AccountService>('accountService');
+      for (const one of persons) {
+        const { _id } = await service.signUp(User.fromJSON(one));
+        one.id = String(_id);
+      }
+    });
+
+    it('[GET] /user/', async () => {
+      const response = await request
+        .get('/user/')
+        .query({
+          search: 'l'
+        })
+        .expect(200);
+      const list: any[] = response.body;
+      list.map((res, idx) => {
+        assert(res.email === persons[idx].email);
+        persons[idx].id = res.id;
+      });
+      return request
+        .get('/user/')
+        .expect(200)
+        .then((response) => {
+          const list: any[] = response.body;
+          list.map((res, idx) => {
+            assert(res.email === persons[idx].email);
+            persons[idx].id = res.id;
```

Comment:
it seems those test cases have to be run in a certain order as they're coupled.
e.g. if you don't have the `GET /user` run first you don't have the `id` on the person entity.

Reply:
Yep. I considered the coverage as most important goal, and I know the order while Mocha run them, so I did that for ease.

If we start to pay attention to test speed, like using concurrency test with Jest, the data initialization should move to common util function, and generate everytime for each case. I'm sure I can do it when it is necessary.